### ISSUE TITLE
row sum and column mean works but row mean gives all NaN, on heterogenous data types

### ISSUE THREAD

**Issue by xuancong84** [pandas-dev/pandas#33202]
When different DataFrame rows are of different types, row sum works but row mean gives all NaN values.

```
df=pd.DataFrame({'a':[1,pd.to_timedelta('1D'),3.], 'b':[4,pd.to_timedelta('1.5D'),6.], 'c':[7,pd.to_timedelta('0.5D'),np.nan]})
# df=pd.DataFrame({'a':[1,2,3.], 'b':[4,5,6.], 'c':[7,8,np.nan]})
display(df.transpose())
display(df.transpose().sum(axis=0))
display(df.transpose().mean(axis=0))
display(df)
display(df.sum(axis=1))
display(df.mean(axis=1))  # bug
```
<img width="332" alt="image" src="https://user-images.githubusercontent.com/10172392/78124015-a4838380-7441-11ea-84fe-9bd45de54f1b.png">

As shown above, the last statement shows the bug by returning all NaN values which is incorrect. Interestingly, if the DataFrame entries are of the same type (uncomment the 2nd line), this bug does not occur.

**Comment by jreback** [comment]
sum of object types generally might work but mean does not. try doing .infer_object() after transpose 

why would you do this in any event? transpose in mixed types is really odd

**Comment by xuancong84** [comment]
@jreback Thanks for your reply! I encounter this because I am doing clinical data statistical analysis. In clinical data, some are floating point, some are integers, some are time durations, etc., so when I compute the average value over some period of time, this bug surfaces.

Currently, transpose=>mean=>transpose-back is my work around, because averaging the column has no problem, only averaging the row has the problem.

**Comment by jreback** [comment]
pandas is column based; mixing dtypes while possible is not recommended; column based dtypes are efficient for both storage and computation

doing numeric operations in heterogeneous data doesn’t make any sense

**Comment by xuancong84** [comment]
So the issue gets closed without getting solved, that is why I have suggested in #6963, "for maximum usage compatibility, we should treat a table as a symmetric rank-2 tensor". Due to low-level CPU optimization mechanical constraint, sum/mean/std/max/min/etc along row direction may incur a performance penalty, this is acceptable, but the operation should not fail and should work consistent with column operations.

Thus, incompetent programming and inherently deficient/defective architectural design does not allow this bug to be solved that easily ^_^
Anyway, you only need to solve this when more people encounter this in their work.
